### PR TITLE
feat: add support for pyarrow arrays as input

### LINKED ISSUE #1415: Support PyArrow arrays as tokenizer input
Most data processing libraries (Datasets, Polars, Pandas, DuckDB, etc.) are integrated with PyArrow, so native (zero-copy if possible) support for PyArrow arrays as input to avoid the unnecessary PyArrow to Python/NumPy conversion (pretty slow for string arrays) would be nice.
 
PS: PyArrow has recently added support for the PyCapsule interface, which should help with the implementation.

### PR DESCRIPTION
**Pull request by notjedi**
closes #1415

### CODE REVIEW

**Comment by notjedi** [comment]
yet to add support for `pyarrow.LargeString`

**Comment by notjedi** [comment]
pulling `arrow` from git because the current stable version `v51.0` links against `v0.20` of `pyo3` while the bindings link against `v0.21`. there is already a merged pull request(apache/arrow-rs#5566) which migrates from `v0.20` to `v0.21` of `pyo3` which i think will be released in `v52.0`

File: bindings/python/src/tokenizer.rs [hunk 1]
```diff
@@ -260,13 +262,43 @@ impl PyAddedToken {
     }
 }
 
+struct PyArrowScalarStringInput<'s>(Cow<'s, str>);
```
Comment (ArthurZucker) [inline bindings/python/src/tokenizer.rs:265]:
```suggestion
[cfg(feature="pyarrow")]
struct PyArrowScalarStringInput<'s>(Cow<'s, str>);
```

**Review by ArthurZucker — COMMENTED** [review]
I think we can leave the dep as optional wdyt?

**Comment by notjedi** [comment]
> I think we can leave the dep as optional wdyt?

makes sense, will get it changed @ArthurZucker

**Comment by notjedi** [comment]
i've added `pyarrow` as a feature in `pyproject.toml`. not sure if that is what we want here. let me know if i need to fix anything else. @ArthurZucker

**Review by ArthurZucker — COMMENTED** [review]
I think we want to avoid having to use git for the repo 😢 
ALso do you have any bench / examples of it being use and fast? 🚀

**Comment by strategy155** [comment]
Good afternoon everyone. How do you think the optimal benchmark should look like? Native pyarrow vs numpy conversion for string arrays?

**Comment by ArthurZucker** [comment]
something like that yeah, I have no idea in what context pyarrow is used as I have not used it, but in optimal context usage

**Comment by strategy155** [comment]
I'm not sure it should be closed. The pr is looking fine.

**Comment by notjedi** [comment]
yes, just update to arrow to the latest stable version `v52.0`

**Pull request closed without merging**